### PR TITLE
[Device Manager] Render other sessions (PSG-668)

### PR DESCRIPTION
**Pull request by onurays**
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Implement other sessions section as in the new layout.

## Motivation and context
#6945 

## Screenshots / GIFs

|Other Sessions|
|-|
|other_sessions|

## Tests

- Enable the new device management feature flag
- Go to "Security & Privacy" settings
- Press the "Show all sessions (V2, WIP)" entry
- See the other sessions are rendered as in the new layout

## Tested devices

- [x] Physical
- [ ] Emulator
- OS version(s):

## Checklist

- [x] Changes has been tested on an Android device or Android emulator with API 21
- [x] UI change has been tested on both light and dark themes
- [x] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [x] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/CurrentSessionView.kt [hunk 1]
```diff
@@ -39,21 +39,14 @@ class CurrentSessionView @JvmOverloads constructor(
         views = ViewCurrentSessionBinding.bind(this)
     }
 
-    fun update(accountCrossSigningIsTrusted: Boolean, legacyMode: Boolean) {
-        renderDeviceType()
-        renderVerificationStatus(accountCrossSigningIsTrusted, legacyMode)
+    fun update(currentDeviceInfo: DeviceFullInfo) {
```
Comment (mnaturel) [inline vector/src/main/java/im/vector/app/features/settings/devices/v2/list/CurrentSessionView.kt:42]: In custom views, it seems we used `render()` to name the method to refresh the views. Maybe we should align to that?

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/CurrentSessionView.kt [hunk 2]
```diff
@@ -39,21 +39,14 @@ class CurrentSessionView @JvmOverloads constructor(
...
+        renderDeviceInfo(currentDeviceInfo.deviceInfo.displayName ?: "")
```
Comment (mnaturel) [inline vector/src/main/java/im/vector/app/features/settings/devices/v2/list/CurrentSessionView.kt:43]: We could use the `currentDeviceInfo.deviceInfo.displayName.orEmpty()` syntax to make things clearer.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsController.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import org.matrix.android.sdk.api.session.crypto.model.RoomEncryptionTrustLevel
+import javax.inject.Inject
+
+class OtherSessionsController @Inject constructor(
+        private val stringProvider: StringProvider,
+        private val dateFormatter: VectorDateFormatter,
+) : TypedEpoxyController<List<DeviceFullInfo>>() {
+
+    override fun buildModels(data: List<DeviceFullInfo>?) {
+        data ?: return
```
Comment (mnaturel) [inline vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsController.kt:35]: The case of `null` is handled later by building a `no result` placeholder. Maybe we should remove this line?

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsView.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+@AndroidEntryPoint
+class OtherSessionsView @JvmOverloads constructor(
+        context: Context,
+        attrs: AttributeSet? = null,
+        defStyleAttr: Int = 0
+) : ConstraintLayout(context, attrs, defStyleAttr) {
+
+    @Inject lateinit var otherSessionsController: OtherSessionsController
+
+    private val views: ViewOtherSessionsBinding
```
Comment (mnaturel) [inline vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsView.kt:38]: We can use the simpler following syntax for view binding:

```
private val views = ViewOtherSessionsBinding.inflate(
    LayoutInflater.from(context),
    this
) 
```
You will need to use a `merge` tag as parent in `view_other_sessions` with `tools:parentTag="androidx.constraintlayout.widget.ConstraintLayout"`.
This avoids uncesseary empty root view in the custom view (right now there is a `ConstraintLayout` inside a `ConstraintLayout`).
The same remark applies for the `CurrentSessionView`.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/SessionDeviceType.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.settings.devices.v2.list
+
+enum class SessionDeviceType {
```
Comment (mnaturel) [inline vector/src/main/java/im/vector/app/features/settings/devices/v2/list/SessionDeviceType.kt:19]: Could we simply name it `DeviceType`?

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/SessionListConstants.kt [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.settings.devices.v2.list
+
+const val NUMBER_OF_OTHER_DEVICES_TO_RENDER = 5
```
Comment (mnaturel) [inline vector/src/main/java/im/vector/app/features/settings/devices/v2/list/SessionListConstants.kt:19]: I would add `internal` modifier to this constant to scope it to the current package.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionItem.kt [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    @EpoxyAttribute
+    var sessionName: String? = null
+
+    @EpoxyAttribute
+    var sessionDescription: String? = null
+
+    override fun bind(holder: Holder) {
+        super.bind(holder)
+        holder.otherSessionDeviceTypeImageView.setImageResource(
```
Comment (mnaturel) [inline vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionItem.kt:46]: I think we should also update the `contentDescription` field using the correct `a11y_` string.

File: vector/src/main/res/layout/view_current_session.xml [hunk 8]
```diff
@@ -21,15 +21,15 @@
         tools:src="@drawable/ic_device_type_mobile" />
 
     <TextView
```
Comment (mnaturel) [inline vector/src/main/res/layout/view_current_session.xml:23]: The size of the view with id `currentSessionVerificationStatusDetailTextView` seems too small compared to Figma. Can you check? I have started a `styles_devices_management.xml` file if needed.

**Review by mnaturel — COMMENTED** [review]
I have added some small comments. When testing it, I have just had a strange behavior: I don't see any display names for all my Android sessions. Is it normal? I am wondering also if we should display something instead of empty space when the name is not known.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsView.kt [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+    init {
+        inflate(context, R.layout.view_other_sessions, this)
+        views = ViewOtherSessionsBinding.bind(this)
+    }
+
+    fun update(devices: List<DeviceFullInfo>) {
+        views.otherSessionsRecyclerView.configureWith(otherSessionsController, hasFixedSize = true)
```
Comment (mnaturel) [inline vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsView.kt:46]: Maybe we need a cleanUp of the `RecyclerView` when the view is view is detached?

File: vector/src/main/res/layout/fragment_settings_devices.xml [hunk 10]
```diff
@@ -1,41 +1,69 @@
...
-        android:id="@+id/waiting_view"
-        layout="@layout/merge_overlay_waiting_view" />
-
-</androidx.constraintlayout.widget.ConstraintLayout>
+    <androidx.constraintlayout.widget.ConstraintLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content">
+
+        <im.vector.app.features.settings.devices.v2.list.DevicesListHeaderView
+            android:id="@+id/deviceListHeaderSectionCurrent"
```
[the remark this answers is not in the document]
Comment (onurays) [reply vector/src/main/res/layout/fragment_settings_devices.xml]: Good idea, done.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/CurrentSessionView.kt (hunk 1 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/settings/devices/v2/list/CurrentSessionView.kt]: Done.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/CurrentSessionView.kt (hunk 2 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/settings/devices/v2/list/CurrentSessionView.kt]: Nice idea, done.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsController.kt (hunk 3 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsController.kt]: Nice catch, done.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsView.kt [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    @Inject lateinit var otherSessionsController: OtherSessionsController
+
+    private val views: ViewOtherSessionsBinding
+
+    init {
+        inflate(context, R.layout.view_other_sessions, this)
+        views = ViewOtherSessionsBinding.bind(this)
+    }
+
+    fun update(devices: List<DeviceFullInfo>) {
```
[the remark this answers is not in the document]
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsView.kt]: Done.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsView.kt (hunk 4 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsView.kt:38]: It couldn't draw background if we use `merge`.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/SessionDeviceType.kt (hunk 5 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/settings/devices/v2/list/SessionDeviceType.kt]: Done.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionItem.kt (hunk 7 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionItem.kt]: Good catch, done.

File: vector/src/main/res/layout/view_current_session.xml (hunk 8 above)
Comment (onurays) [reply vector/src/main/res/layout/view_current_session.xml:23]: Right, fixed.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/SessionListConstants.kt (hunk 6 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/settings/devices/v2/list/SessionListConstants.kt]: Done.

File: vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsView.kt (hunk 9 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/settings/devices/v2/list/OtherSessionsView.kt:47]: Done. I will raise the empty device name issue in the internal room.

**Pull request merged**